cmdstanpy/model.py

```diff
@@ -652,91 +677,130 @@ def _read_progress(self, proc: subprocess.Popen, pbar: List[Any]) -> bytes:
+                            pbar_sampling.update(count)
+
+            # read and process rest of the stdout if needed
+            warmup_cumulative_count = 0
+            sampling_cumulative_count = 0
+            for output in proc.stdout:
+                stdout += output
+                # decode bytes to string
+                output = output.decode('utf-8').strip()
+                if output.startswith('Iteration'):
+                    # run regular expression
+                    match = re.search(pattern_compiled, output)
+                    if match:
+
+                        # check if pbars need reset
+                        if num_warmup is None or num_sampling is None:
+                            total_count = int(match.group(2))
+                            if num_warmup is None and num_sampling is None:
+                                num_warmup = total_count // 2
+                                num_sampling = total_count - num_warmup
+                                pbar_warmup.total = num_warmup
+                                pbar_sampling.total = num_sampling
+                            elif num_warmup is None:
+                                num_warmup = total_count - num_sampling
+                                pbar_warmup.total = num_warmup
+                            else:
+                                num_sampling = total_count - num_warmup
+                                pbar_sampling.total = num_sampling
+
+                        # raw_count = warmup + sampling
+                        raw_count = int(match.group(1))
+                        if match.group(3).lower() == 'warmup':
+                            count, count_warmup = (
+                                # steps
+                                raw_count - count_warmup,
+                                # cumulative count
+                                raw_count,
+                            )
+                            # increment warmup_cumcount by 'count'
+                            warmup_cumulative_count += count
+                        elif match.group(3).lower() == 'sampling':
+                            count, count_sampling = (
+                                # steps
+                                raw_count - num_warmup - count_sampling,
+                                # cumulative count
+                                raw_count - num_warmup,
+                            )
+                            # increment progressbar by 'count'
+                            sampling_cumulative_count += count
+            # update warmup pbar if needed
+            if warmup_cumulative_count:
+                pbar_warmup.update(warmup_cumulative_count)
+                pbar_warmup.refresh()
+            # update sampling pbar if needed
+            if sampling_cumulative_count:
+                pbar_sampling.update(sampling_cumulative_count)
+                pbar_sampling.refresh()
+
```

Comment:
also true for comments here, excepting comments about warmup vs. sampling lines #725-#726